R/convert-gene-ids.R

```diff
@@ -0,0 +1,125 @@
+#' symbols. If any gene ids are not found or if the gene symbol is not defined,
+#' the input gene id is returned, unless the `leave_na` is set to `TRUE`.
+#'
+#'
+#' @param ensembl_ids A character vector of Ensembl gene ids to translate to
+#'  gene symbols.
+#' @param sce A SingleCellExperiment object containing gene ids and gene symbols
+#'  to use for translation.
+#' @param leave_na logical indicating whether to leave NA values in the output.
+#'  Default is `FALSE`
+#'
+#' @return A vector of gene symbols corresponding to the input Ensembl ids.
+#' @export
+#'
+#' @import SingleCellExperiment
+#'
+#' @examples
+#' \dontrun{
+#' # convert a set of Ensembl ids to gene symbols
+#' # using a SingleCellExperiment reference
+#' ensembl_ids <- c("ENSG00000141510", "ENSG00000134323")
+#' gene_symbols <- ensembl_to_symbol(ensembl_ids, sce)
+#' gene_symbols
+#' ### [1] "TP53" "MYCN"
+#' }
+ensembl_to_symbol <- function(ensembl_ids, sce, leave_na = FALSE) {
+  stopifnot(
+    "`sce` must be a SingleCellExperiment object ." = is(sce, "SingleCellExperiment"),
+    "`ensembl_ids` must be a character vector." = is.character(ensembl_ids),
+    "`sce` must contain both a `gene_ids` and `gene_symbol` column in the row data." =
+      all(c("gene_ids", "gene_symbol") %in% names(rowData(sce))),
+    "`leave_na` must be TRUE or FALSE." = is.logical(leave_na)
+  )
+
+  id_match <- match(ensembl_ids, rowData(sce)$gene_ids)
+  gene_symbols <- rowData(sce)[id_match, "gene_symbol"]
+
+  missing_symbols <- is.na(gene_symbols)
+  if (!leave_na && any(missing_symbols)) {
+    warning("Not all `ensembl_ids` are present in sce `gene_ids`, using input ids for missing values.")
+    gene_symbols[missing_symbols] <- ensembl_ids[missing_symbols]
+  }
+
+  return(gene_symbols)
+}
+
+#' Set the row names of an ScPCA SingleCellExperiment object to gene symbols
+#'
+#' The SingleCellExperiment objects produced as part of ScPCA are indexed by
+#' Ensembl gene ids, as those are more stable than gene symbols. However,
+#' for many applications gene symbols are useful. This function converts the
+#' row names (indexes) of a SingleCellExperiment object to gene symbols based.
+#'
+#' Internal data structures such as the list of highly variable genes and the
+#' rotation matrix for the PCA are also updated to use gene symbols, if present
+#' (and not disabled by the `convert_hvg` and `convert_pca` arguments).
+#'
+#' Note that using this function will result in non-unique row ids as no
+#' de-duplication is currently performed.
+#'
+#' @param sce A SingleCellExperiment object containing gene ids and gene symbols.
+#' @param convert_hvg Logical indicating whether to convert highly variable genes to gene symbols.
+#' @param convert_pca Logical indicating whether to convert PCA rotation matrix to gene symbols.
+#'
+#' @return A SingleCellExperiment object with row names set as gene symbols.
+#' @export
+#'
+#' @import SingleCellExperiment
```

Comment:
```suggestion
#' @import SingleCellExperiment
#' @import S4Vectors
```

While the tests for this pass, I was unable to run this function locally:
```
Error: 'metadata' is not an exported object from 'namespace:SingleCellExperiment'
```

Turns out this comes from `S4Vectors`!

Reply:
I'm not sure how you are doing the local tests, but it should pass if you load `SingleCellExperiment`, or if if you actually load the package. The following works in a fresh R session:

```
> library(rOpenScPCA)
> sce <- readRDS(testthat::test_path("data", "scpca_sce.rds"))
> sce2 <- sce_to_symbols(sce)
```

As does:

```
> sce <- readRDS(testthat::test_path("data", "scpca_sce.rds"))
> sce2 <- rOpeScPCA::sce_to_symbols(sce)
```

Reply:
I loaded with `devtools::load_all()`.  The [docs](https://devtools.r-lib.org/reference/load_all.html#differences-with-loadnamespace-and-library-) suggest that `load_all` should actually load up more than `library`, so the fact that this doesn't suffice surprises/confuses me:

> Whereas [loadNamespace()](https://rdrr.io/r/base/ns-load.html) and [library()](https://rdrr.io/r/base/library.html) only load package dependencies when they are needed, load_all() loads all packages referenced in Imports at load time.

It works as expected when explicitly load `library(SingleCellExperiment).


In a _separate fresh_ R session without any sneaky renv business, I still got the error:

```
> remotes::install_github("AlexsLemonade/rOpenScPCA@fe08c631cf61437710b5bac8155978c5bf9a2bea")

######### I quit and restarted R here ##########

> library(rOpenScPCA)
> sce <- readRDS("~/ALSF/open-scpca/rOpenScPCA/tests/testthat/data/scpca_sce.rds")
> converted_sce <- sce_to_symbols(sce)
Error in metadata(sce) : could not find function "metadata"
In addition: Warning message:
In sce_to_symbols(sce) :
  Not all rows have gene symbols, using ensembl ids for missing values.
```

Reply:
More context:

```
> library(rOpenScPCA)
> sce <- readRDS("~/ALSF/open-scpca/rOpenScPCA/tests/testthat/data/scpca_sce.rds")
> converted_sce <- sce_to_symbols(sce)
Error in metadata(sce) : could not find function "metadata"
In addition: Warning message:
In sce_to_symbols(sce) :
  Not all rows have gene symbols, using ensembl ids for missing values.
>
>
>
> sessionInfo()
R version 4.4.0 (2024-04-24)
Platform: aarch64-apple-darwin20
Running under: macOS 15.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] rOpenScPCA_0.1.0

loaded via a namespace (and not attached):
 [1] GenomeInfoDb_1.40.1         R6_2.5.1                   
 [3] SparseArray_1.4.8           zlibbioc_1.50.0            
 [5] Matrix_1.7-1                lattice_0.22-6             
 [7] MatrixGenerics_1.16.0       matrixStats_1.4.1          
 [9] abind_1.4-8                 GenomeInfoDbData_1.2.12    
[11] S4Arrays_1.4.1              XVector_0.44.0             
[13] Biobase_2.64.0              BiocGenerics_0.50.0        
[15] UCSC.utils_1.0.0            stats4_4.4.0               
[17] GenomicRanges_1.56.2        IRanges_2.38.1             
[19] grid_4.4.0                  DelayedArray_0.30.1        
[21] compiler_4.4.0              httr_1.4.7                 
[23] tools_4.4.0                 crayon_1.5.3               
[25] jsonlite_1.8.9              S4Vectors_0.42.1           
[27] SummarizedExperiment_1.34.0 SingleCellExperiment_1.28.1
```

Reply:
That seems to indicate that `S4Vectors` is loaded. I'm not sure what is going wrong.

Reply:
Tests are probably passing b/c of

Reply:
S tests are probably only passing because: